### DISCUSSION TITLE
Error in calculating the angle of attack

### DISCUSSION THREAD

**Discussion by 4ayokk** [JSBSim-Team/jsbsim#979] [Q&A]
Hello. I see flight dynamics not working properly in jsbsim. I have checked Cessna, Boeing 737. There is an error with the angle of attack (alpha) during flight everywhere. 
And the error is not in the configuration file! The error is in the flight dynamics calculations themselves!
When I deflect the elevator for climb and leave it in this position for a few seconds, I see:
1. Initially, the angle of attack increases correctly to the value specified in the configuration file (see the figure below)
2. Then the angle of attack starts to increase again to higher values. Although Nz starts to decrease (see the figure). When Nz decreases, the angle of attack cannot increase!

I performed flights without rolls and yaw. 

## Boeing:
![boeing](https://github.com/JSBSim-Team/jsbsim/assets/113011334/cadbe6fd-2a8c-4461-b08a-9841fea6c951)
## Cessna:
![Cessna_alpha](https://github.com/JSBSim-Team/jsbsim/assets/113011334/4582cbcb-f51c-4066-aea8-fffe124eca5a)

Can you please help me? Is it possible to modify jsbsim and fix this error in calculating the angle of attack.

Below is an example of how the angle of attack should behave relative to Nz.
![right_alpha](https://github.com/JSBSim-Team/jsbsim/assets/113011334/0d83c7d7-520c-40a5-895f-56372c3aebf5)

**Comment by seanmcleod** [comment]
I think you might be jumping to conclusions to say that the JSBSim flight dynamics calculations are wrong :wink: 

At first glance, glancing at your graphs for the Boeing case, it looks like to some extent you've emulated the Air France AF447 trajectory, in terms of keeping the nose up, landing up in a deep stall with a large descent rate.

> Initially, the angle of attack increases correctly to the value specified in the configuration file

What value in the configuration file are you referring to? Are you referring to the alpha value for maximum Cl for the 737 model in JSBSim, i.e. this graph?

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/c8074698-2a92-4edd-856d-df0b86cc6df0)

[AF447 Accident Report](https://bea.aero/docspa/2009/f-cp090601.en/pdf/f-cp090601.en.pdf)

Take a look at bullet points on the right hand edge of the FDR graphs, a vertical speed of -10,000ft/min, pitch attitude of +15deg and an AoA of 41.5deg.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/edde7103-94dd-4d18-84b7-bd6d9b7f2673)

> When Nz decreases, the angle of attack cannot increase!

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/2fe5c0dc-08e5-4df3-8b14-d73105693c94)

Take the TAS and the vertical speed and calculate the flight path angle, and compare it to the flight path angle calculated by JSBSim, and compare it to the pitch attitude to see whether the AoA makes sense.

**Comment by 4ayokk** [comment]
I did another experiment, and the elevator deviation is much smaller. But the AoA still continues to grow as nz falls, even though we believe it should fall.

![boeing737](https://github.com/JSBSim-Team/jsbsim/assets/113011334/75450f1b-e241-42c1-8714-de66063a8117)

**Comment by seanmcleod** [comment]
Are you claiming it's impossible for AoA to increase while Nz is decreasing? Under all circumstances or just in this sort of maneuver? Explain why.

Also note in terms of the increase in AoA, in this example the AoA only increases from ~5.45deg to ~5.85deg, so only about 0.4deg.

Are you familiar with the Phugoid mode? Where the aircraft climbs and descends, trading potential energy for kinetic energy while maintaining roughly the same AoA.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/c16d7795-45e0-40db-baea-f2fc5ffe5906)

**Comment by 4ayokk** [comment]
If you conditionally divide the mode into the first and second parts, you can see the following:
First part {1} - when the control signal appears, the AoA initially behaves correctly. The angle increases up to a certain value, and then begins to decrease (o).
The increase in angle is due to the moments from moving the elevator.

Second part {2} - The AoA stops decreasing and starts increasing. It is not clear what causes the angle of attack to increase in the second stage.

Note that the value by which AoA increases in the second stage is quite large compared to the increase in the first stage. For example, the increase in AoA at the second stage is 0.4 degrees, while at the first stage it is 0.55. (Figure 1)


![Boeing_1](https://github.com/JSBSim-Team/jsbsim/assets/113011334/ccc99b93-524b-4522-8188-02a8e208f4df)
---
![Cessna](https://github.com/JSBSim-Team/jsbsim/assets/113011334/74386585-e048-4e8e-8968-dce6ea506596)
---
![Boeing_2](https://github.com/JSBSim-Team/jsbsim/assets/113011334/708f4d8b-e1de-411a-9863-f8874ae56ebc)

**Comment by agodemar** [comment]
@4ayokk what is your background? 
Are you aware of the various basic phenomena of flight dynamics?
Did you check in your tests and models if the thrust axis is barycentric or not? In which case the pitching moment inbalance determining AoA changes would be due to thrust and thrust variations with altitude.

**Comment by seanmcleod** [comment]
@4ayokk plot the velocities `u` and `w`, AoA is the arctan between them:

https://github.com/JSBSim-Team/jsbsim/blob/ae4652a268d0475dbf93d47a32ab7ddf4120aaef/src/models/FGAuxiliary.cpp#L161

Also plot flight path angle (gamma) and pitch attitude (theta) since AoA is also the difference between the two.

I think what's contributing is a larger decrease in `u` than you're taking into account.

**Comment by seanmcleod** [comment]
@4ayokk I ran the following test using the JSBSim 737 model.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/db790b57-4f6a-4f41-9f66-0ad361273505)

So trimmed at 250KIAS at 10,000ft, 1s after trim a small elevator step input is applied.

- Trim Pitch Attitude: 3.25deg
- Steady State Pitch Attitude: 4.0deg

- Trim Flight Path Angle: 0deg
- Steady State Flight Path Angle: 0.22deg

- Trim AoA: 3.25deg
- Steady State AoA: 3.82deg

Looks like a very lightly damped Phugoid.

P.S Interesting effect depending on minimum drag speed 
https://aviation.stackexchange.com/questions/98099/after-the-phugoid-mode-subsides-in-the-steady-state-why-is-there-a-small-incre

**Comment by seanmcleod** [comment]
Here you can see how Nz is driven by the airspeed given the fairly constant AoA that you see above after the initial pitch up.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/6a2bfb96-3545-4451-a965-a666026a84f4)

Zoomed in for the first 45s.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/8330fad2-c416-42cf-af96-5baad675bec9)

**Reply by agodemar** [reply]
Typical Phugoid

**Comment by 4ayokk** [comment]
@agodemar thank you! 

> Did you check in your tests and models if the thrust axis is barycentric or not?

As it turned out, the problem was that I had the wrong shoulder for the thrust vector, which led to this behaviour of the angle of attack.

**Comment by 4ayokk** [comment]
I want to express my sincere gratitude for Your help. Your words and professional guidance helped us find the error that had been bothering us for a long time. You turned out to be real heroes who helped our project move forward.

**Comment by 4ayokk** [comment]
At the same time, I would like to express my deepest apologies. I made a mistake in my dubious judgement and JSBSim actually counts correctly. You were right, and I am grateful for your patience, understanding, and support.

**Comment by seanmcleod** [comment]
No problem. Out of interest can you tell us a bit about your project? Always interested to know what JSBSim is being used for.

> The error is in the flight dynamics calculations themselves!

Generally better to go with a question mark, rather than an exclamation mark, unless you're 100% certain! :wink:

**Comment by 4ayokk** [comment]
Well, I recently got a job at an aviation company that develops flight simulators. But I'm a radio engineer by education, and I know very little about aerodynamics. So, in order to not only read books, but also to see how it all works in practice, and at the same time to explore how such models work, I decided to try to create my own aircraft model based on JSBSim. At work, we have our own models and theoretical calculations for them, and my task was to make a model that would coincide with these calculations. When I chose FDM, I really liked the realism of JSBSim and the fact that it is used by NASA in its work.

**Reply by agodemar** [reply]
@4ayokk you should read (study) books at some point. JSBSim, like any other 6-DoF flight sim library, could become a complex toy to play with, unless you know what you are doing. To know, you have to upgrade your knowledge on the basic phenomena in flight dynamics
